### PR TITLE
[KBFS-1150] Split MDServerLocal into MDServer{Memory,Disk}

### PR DESCRIPTION
**Pull request by akalin-keybase**
Rewrite MDServerMemory to use plain maps.

Leave MDServerDisk mostly alone for now, but
try to decomp out some shared code.

### CODE REVIEW

**Comment by akalin-keybase** [comment]
Note that this depends on KBFS-1159, so don't look at this until KBFS-1159 is merged.

**Comment by akalin-keybase** [comment]
Okay, this is ready for review.

File: libkbfs/mdserver_disk.go [hunk 1]
```diff
@@ -100,70 +76,55 @@ func NewMDServerLocal(config Config, handleDbfile string, mdDbfile string,
...
-// GetForHandle implements the MDServer interface for MDServerLocal.
-func (md *MDServerLocal) GetForHandle(ctx context.Context, handle BareTlfHandle,
+// GetForHandle implements the MDServer interface for MDServerDisk.
+func (md *MDServerDisk) GetForHandle(ctx context.Context, handle BareTlfHandle,
 	mStatus MergeStatus) (TlfID, *RootMetadataSigned, error) {
 	id := NullTlfID
-	md.shutdownLock.RLock()
-	defer md.shutdownLock.RUnlock()
-	if *md.shutdown {
+	if md.isShutdown() {
```
Comment (strib) [inline libkbfs/mdserver_disk.go]: This (and other functions below) no longer holds the shutdown lock throughout the call, so a shutdown could sneak in right after this check and destroy the storage, right?

File: libkbfs/mdserver_memory.go [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+	deviceKID keybase1.KID
+}
+
+type mdBlockMem struct {
+	// An encoded RootMetdataSigned.
+	encodedMd []byte
+	timestamp time.Time
+}
+
+// MDServerMemory just stores blocks in local leveldb instances.
```
Comment (strib) [inline libkbfs/mdserver_memory.go]: No more leveldb in this file.

File: libkbfs/mdserver_memory.go [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+	if err != nil {
+		return id, nil, MDServerError{err}
+	}
+	return id, nil, nil
+}
+
+// GetForTLF implements the MDServer interface for MDServerMemory.
+func (md *MDServerMemory) GetForTLF(ctx context.Context, id TlfID,
+	bid BranchID, mStatus MergeStatus) (*RootMetadataSigned, error) {
+	if md.isShutdown() {
```
Comment (strib) [inline libkbfs/mdserver_memory.go]: Same comment here about the shutdown mutex.

File: libkbfs/mdserver_memory.go [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+// GetRange implements the MDServer interface for MDServerMemory.
+func (md *MDServerMemory) GetRange(ctx context.Context, id TlfID,
+	bid BranchID, mStatus MergeStatus, start, stop MetadataRevision) (
+	[]*RootMetadataSigned, error) {
+	md.log.CDebugf(ctx, "GetRange %d %d (%s)", start, stop, mStatus)
+	if md.isShutdown() {
+		return nil, errors.New("MD server already shut down")
+	}
+
+	if mStatus == Merged && bid != NullBranchID {
```
Comment (strib) [inline libkbfs/mdserver_memory.go]: Could factor out a good chunk of this function and share with `GetForTLF`, right?

File: libkbfs/mdserver_memory.go [hunk 5]
```diff
@@ -0,0 +1,641 @@
...
+
+	mStatus := rmds.MD.MergedStatus()
+	bid := rmds.MD.BID
+
+	if mStatus == Merged {
+		if bid != NullBranchID {
+			return MDServerErrorBadRequest{Reason: "Invalid branch ID"}
+		}
+	} else {
+		if bid == NullBranchID {
```
Comment (strib) [inline libkbfs/mdserver_memory.go]: Could save an indent by combining this with the `else`.

File: libkbfs/mdserver_remote.go [hunk 6]
```diff
@@ -90,6 +92,11 @@ func NewMDServerRemote(config Config, srvAddr string, ctx Context) *MDServerRemo
 	return mdServer
 }
 
+// RemoteAddress returns the remote bserver this client is talking to
```
Comment (strib) [inline libkbfs/mdserver_remote.go]: "bserver" -> "mdserver"?

File: libkbfs/root_metadata.go [hunk 7]
```diff
@@ -459,6 +460,40 @@ func (md *RootMetadata) CheckValidSuccessor(
 	return nil
 }
 
+// CheckValidSuccessorForServer is like CheckValidSuccessor but with
+// server-specific error messages and extra checks.
```
Comment (strib) [inline libkbfs/root_metadata.go]: I don't see any "extra checks".

**Comment by strib** [comment]
Looks good mod nits and another look at the shutdown locking.  Thanks!

File: libkbfs/mdserver_disk.go (hunk 1 above)
Comment (akalin-keybase) [inline libkbfs/mdserver_disk.go]: Ugh, yeah, you're right. Reverted all those changes; it'll all be different in the follow-up PR anyway.

File: libkbfs/mdserver_memory.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/mdserver_memory.go]: Fixed.

File: libkbfs/mdserver_memory.go (hunk 3 above)
Comment (akalin-keybase) [inline libkbfs/mdserver_memory.go]: Yeap, reworked to have a single lock for everything.

File: libkbfs/mdserver_memory.go (hunk 4 above)
Comment (akalin-keybase) [inline libkbfs/mdserver_memory.go]: Done. Will do so for `MDServerDisk` in the next PR.

File: libkbfs/mdserver_memory.go (hunk 5 above)
Comment (akalin-keybase) [inline libkbfs/mdserver_memory.go]: Done.

File: libkbfs/mdserver_remote.go (hunk 6 above)
Comment (akalin-keybase) [inline libkbfs/mdserver_remote.go]: Done.

File: libkbfs/root_metadata.go (hunk 7 above)
Comment (akalin-keybase) [inline libkbfs/root_metadata.go]: Removed.

**Pull request merged**